daemon/volumes.go

```diff
@@ -65,9 +65,21 @@ func (m mounts) parts(i int) int {
 // 2. Select the volumes mounted from another containers. Overrides previously configured mount point destination.
 // 3. Select the bind mounts set by the client. Overrides previously configured mount point destinations.
 // 4. Cleanup old volumes that are about to be reassigned.
-func (daemon *Daemon) registerMountPoints(container *container.Container, hostConfig *containertypes.HostConfig) error {
+func (daemon *Daemon) registerMountPoints(container *container.Container, hostConfig *containertypes.HostConfig) (retErr error) {
 	binds := map[string]bool{}
 	mountPoints := map[string]*volume.MountPoint{}
+	defer func() {
+		// clean up the container mountpoints once return with error
+		if retErr != nil && len(mountPoints) != 0 {
```

Comment:
no need to check `len` on `mountPoints` since it's a `range`

I also think this may be too deep in the stack. I'd expect the caller to unwind this on error (and use https://github.com/docker/docker/blob/master/daemon/mounts.go#L20)